File: node/orca-agent/tests/utils/error_handler.py
import enum
import json
import traceback
import uuid
from datetime import datetime
from typing import Dict, Any, Optional, Callable, TypeVar, Union

class ErrorCategory(enum.Enum):
    """Standardized Error Categories"""
    CONFIGURATION = 'CONFIGURATION'
    NETWORK = 'NETWORK'
    DEPENDENCY = 'DEPENDENCY'
    RUNTIME = 'RUNTIME'
    VALIDATION = 'VALIDATION'
    INTEGRATION = 'INTEGRATION'
    PERMISSION = 'PERMISSION'

class ErrorSeverity(enum.Enum):
    """Standardized Error Severity Levels"""
    INFO = 'INFO'
    LOW = 'LOW'
    MEDIUM = 'MEDIUM'
    HIGH = 'HIGH'
    CRITICAL = 'CRITICAL'

class TestScriptError(Exception):
    """Standardized Error Class for Test Scripts"""
    
    def __init__(
        self, 
        message: str, 
        category: ErrorCategory = ErrorCategory.RUNTIME,
        severity: ErrorSeverity = ErrorSeverity.MEDIUM,
        context: Optional[Dict[str, Any]] = None,
        source: Optional[str] = None
    ):
        super().__init__(message)
        self.id = str(uuid.uuid4())
        self.category = category
        self.severity = severity
        self.context = context or {}
        self.source = source or 'TestScript'
        self.timestamp = datetime.now().timestamp()
        self.stack_trace = traceback.format_exc()

    def to_structured_error(self) -> Dict[str, Any]:
        """Convert error to a standardized structured error format"""
        return {
            'id': self.id,
            'message': str(self),
            'category': self.category.value,
            'severity': self.severity.value,
            'timestamp': int(self.timestamp),
            'context': self.context,
            'stackTrace': self.stack_trace,
            'source': self.source
        }

    def log(self, custom_logger: Optional[Callable[[Dict[str, Any]], None]] = None):
        """Log the error with optional custom logging function"""
        structured_error = self.to_structured_error()
        
        if custom_logger:
            custom_logger(structured_error)
        else:
            print(json.dumps(structured_error, indent=2))

class ErrorHandler:
    """Comprehensive Error Handling Utility"""
    
    @staticmethod
    def create(
        message: str, 
        category: ErrorCategory = ErrorCategory.RUNTIME,
        severity: ErrorSeverity = ErrorSeverity.MEDIUM,
        context: Optional[Dict[str, Any]] = None,
        source: Optional[str] = None
    ) -> TestScriptError:
        """Create a standardized error"""
        return TestScriptError(
            message, 
            category=category, 
            severity=severity, 
            context=context,
            source=source
        )

    @staticmethod
    async def safe_execute(
        fn: Callable[[], Any], 
        error_handler: Optional[Callable[[TestScriptError], None]] = None,
        fallback_value: Optional[Any] = None
    ) -> Optional[Any]:
        """Safely execute a function with error handling"""
        try:
            return await fn()
        except Exception as error:
            if isinstance(error, TestScriptError):
                test_error = error
            else:
                test_error = ErrorHandler.create(
                    str(error), 
                    category=ErrorCategory.RUNTIME, 
                    severity=ErrorSeverity.HIGH
                )
            
            if error_handler:
                error_handler(test_error)
            else:
                test_error.log()
            
            return fallback_value

    @staticmethod
    def log_entry(
        error: Union[TestScriptError, Exception, str],
        category: Optional[ErrorCategory] = None,
        severity: Optional[ErrorSeverity] = None,
        context: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Create a log entry for tracking errors"""
        if isinstance(error, TestScriptError):
            structured_error = error.to_structured_error()
        elif isinstance(error, Exception):
            test_error = ErrorHandler.create(
                str(error), 
                category=category or ErrorCategory.RUNTIME,
                severity=severity or ErrorSeverity.MEDIUM,
                context=context
            )
            structured_error = test_error.to_structured_error()
        else:
            test_error = ErrorHandler.create(
                error, 
                category=category or ErrorCategory.RUNTIME,
                severity=severity or ErrorSeverity.MEDIUM,
                context=context
            )
            structured_error = test_error.to_structured_error()

        return structured_error

File: node/tests/utils/error_handler.ts
import { v4 as uuidv4 } from 'uuid';
import { ErrorCategory, ErrorSeverity, StructuredError } from './error_constants';

export class TestScriptError extends Error {
  public readonly id: string;
  public readonly category: ErrorCategory;
  public readonly severity: ErrorSeverity;
  public readonly context?: Record<string, any>;
  public readonly timestamp: number;
  public readonly source?: string;

  constructor(
    message: string,
    options: {
      category?: ErrorCategory;
      severity?: ErrorSeverity;
      context?: Record<string, any>;
      source?: string;
    } = {}
  ) {
    super(message);
    this.name = 'TestScriptError';
    this.id = uuidv4();
    this.category = options.category || ErrorCategory.RUNTIME;
    this.severity = options.severity || ErrorSeverity.MEDIUM;
    this.context = options.context;
    this.source = options.source || 'TestScript';
    this.timestamp = Date.now();
  }

  /**
   * Convert error to a standardized structured error format
   */
  toStructuredError(): StructuredError {
    return {
      id: this.id,
      message: this.message,
      category: this.category,
      severity: this.severity,
      timestamp: this.timestamp,
      context: this.context,
      stackTrace: this.stack,
      source: this.source
    };
  }

  /**
   * Log the error with optional custom logging function
   */
  log(customLogger?: (error: StructuredError) => void): void {
    const structuredError = this.toStructuredError();
    
    if (customLogger) {
      customLogger(structuredError);
      return;
    }

    console.error(JSON.stringify(structuredError, null, 2));
  }
}

export class ErrorHandler {
  /**
   * Create a standardized error
   */
  static create(
    message: string,
    options: {
      category?: ErrorCategory;
      severity?: ErrorSeverity;
      context?: Record<string, any>;
      source?: string;
    } = {}
  ): TestScriptError {
    return new TestScriptError(message, options);
  }

  /**
   * Safely execute a function with error handling
   */
  static async safeExecute<T>(
    fn: () => Promise<T>,
    options: {
      errorHandler?: (error: TestScriptError) => void;
      fallbackValue?: T;
    } = {}
  ): Promise<T | undefined> {
    try {
      return await fn();
    } catch (error) {
      const testError = error instanceof TestScriptError 
        ? error 
        : this.create(
            error instanceof Error ? error.message : 'Unknown error', 
            { 
              category: ErrorCategory.RUNTIME, 
              severity: ErrorSeverity.HIGH 
            }
          );
      
      if (options.errorHandler) {
        options.errorHandler(testError);
      } else {
        testError.log();
      }
      
      return options.fallbackValue;
    }
  }

  /**
   * Create a log entry for tracking errors
   */
  static logEntry(
    error: TestScriptError | Error | string,
    options: {
      category?: ErrorCategory;
      severity?: ErrorSeverity;
      context?: Record<string, any>;
    } = {}
  ): StructuredError {
    let structuredError: StructuredError;

    if (error instanceof TestScriptError) {
      structuredError = error.toStructuredError();
    } else if (error instanceof Error) {
      const testError = this.create(error.message, {
        category: options.category,
        severity: options.severity,
        context: options.context
      });
      structuredError = testError.toStructuredError();
    } else {
      const testError = this.create(error, {
        category: options.category,
        severity: options.severity,
        context: options.context
      });
      structuredError = testError.toStructuredError();
    }

    return structuredError;
  }
}
